icu4c/source/common/locid.cpp

```diff
@@ -302,10 +302,14 @@ struct Locale::Heap : public UMemory {
         copyToArray<&Heap::region>(region, this);
     }
 
-    Heap(const Heap& other)
+    Heap(const Heap& other, UErrorCode& status)
         : variantBegin(other.variantBegin),
           fullName(other.fullName),
           baseName(other.baseName) {
+        if (fullName.isEmpty() != other.fullName.isEmpty() ||
+            baseName.isEmpty() != other.baseName.isEmpty()) {
+            status = U_MEMORY_ALLOCATION_ERROR;
```

Comment:
Don't set a failure code if there is one already.

Suggestion:
- Don't copy-construct the names in the initializer
- First line here: check the status, do nothing if failure
- Then copy the names
- Then check if either copy failed and set the error

Alternative: Add a UErrorCode to the copy constructor, have that not copy when already failure, else set a failure if OOM. Then you can copy-construct in the initializers and need not check later that the copies worked. Simpler & safer.

Reply:
Done.